cpp/include/raft/core/mdarray.hpp

```diff
@@ -677,12 +703,90 @@ auto make_host_vector_view(ElementType* ptr, size_t n)
  * @return raft::device_vector_view
  */
 template <typename ElementType, typename LayoutPolicy = layout_c_contiguous>
-auto make_device_vector_view(ElementType* ptr, size_t n)
+inline auto make_device_vector_view(ElementType* ptr, size_t n)
 {
-  detail::vector_extent extents{n};
+  vector_extent extents{n};
   return device_vector_view<ElementType, LayoutPolicy>{ptr, extents};
 }
 
+template <typename... Extents>
+using ensure_integral_extents =
+  std::enable_if_t<(true && ... && std::is_integral_v<Extents>), void>;
+
+/**
+ * @brief Create a host mdarray.
+ * @tparam ElementType the data type of the matrix elements
+ * @tparam LayoutPolicy policy for strides and layout ordering
+ * @param exts dimensionality of the array (series of integers)
+ * @return raft::host_mdarray
+ */
+template <typename ElementType,
+          typename LayoutPolicy = layout_c_contiguous,
+          typename... Extents,
+          typename = ensure_integral_extents<Extents...>>
+inline auto make_host_mdarray(Extents... exts)
+{
+  using extent_t  = extents<((void)exts, dynamic_extent)...>;
+  using mdarray_t = host_mdarray<ElementType, extent_t, LayoutPolicy>;
+
+  typename mdarray_t::extents_type extent{exts...};
+  typename mdarray_t::mapping_type layout{extent};
+  typename mdarray_t::container_policy_type policy;
+
+  return mdarray_t{layout, policy};
+}
+
+/**
+ * @brief Create a device mdarray.
+ * @tparam ElementType the data type of the matrix elements
+ * @tparam LayoutPolicy policy for strides and layout ordering
+ * @param stream cuda stream for ordering events
+ * @param exts dimensionality of the array (series of integers)
+ * @return raft::device_mdarray
+ */
+template <typename ElementType,
+          typename LayoutPolicy = layout_c_contiguous,
+          typename... Extents,
+          typename = ensure_integral_extents<Extents...>>
+inline auto make_device_mdarray(rmm::cuda_stream_view stream, Extents... exts)
```

Comment:
I think I started the wrong convention here (sorry ...), `raft::handle_t` might be better than stream. We will need functions like `zeros` `ones` and these functions might use thrust policy to fill the array.

Reply:
I wouldn't like to do so, because there are still places in our codebase where the `stream` is available, but `handle_t` isn't (and we can go one direction, but not the other). There are also places where both a `handle_t` and a `stream` provided as arguments and there is no guarantee the provided stream is the same as the one in the handle, which can lead to thurst functions running out of order w.r.t. the other code. Also, when `handle_t` is not available, one can use `rmm::exec_policy(stream)` for thrust functions...

Anyway, as you suggested, I believe it's a wrong place for this conversation, because the scope of it is much bigger than this PR. I'd just like to merge these couple helpers so I can remove them from the other PR :)

Reply:
Got it. Thank you for the explanation.